Prevent mqueue from implicitely becoming a bind mount with --ipc=host

mlaventure opened the pull request:
Currently, when running a container with `--ipc=host`, if `/dev/mqueue` is
a standard directory on the hos the daemon will bind mount it allowing
the container to create/modify files on the host.

This commit forces `/dev/mqueue` to always be of type `mqueue` except when
the user explicitely requested something to be bind mounted to
`/dev/mqueue`.

Signed-off-by: Kenfe-Mickael Laventure mickael.laventure@gmail.com

---

ping @crosbymichael @rhatdan

mlaventure committed:
Prevent mqueue from implicitely becoming a bind mount with --ipc=host

Currently, when running a container with --ipc=host, if /dev/mqueue is
a standard directory on the hos the daemon will bind mount it allowing
the container to create/modify files on the host.

This commit forces /dev/mqueue to always be of type mqueue except when
the user explicitely requested something to be bind mounted to
/dev/mqueue.

Signed-off-by: Kenfe-Mickael Laventure <mickael.laventure@gmail.com>

mlaventure commented:
Linking #19876

GordonTheTurtle commented:
Job: Docker-PRs-userns [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-userns/6012/console):

~~~console

---
pping root)
SKIP: docker_cli_run_test.go:3667: DockerSuite.TestRunModeNetContainerHostname (Test cannot be run when remapping root)
SKIP: docker_cli_run_test.go:2479: DockerSuite.TestRunModePidHost (Test cannot be run when remapping root)
PASS: docker_cli_run_test.go:2501: DockerSuite.TestRunModeUTSHost	0.848s
SKIP: docker_cli_run_test.go:2102: DockerSuite.TestRunMountOrdering (Test cannot be run when remapping root)

----------------------------------------------------------------------
FAIL: docker_cli_run_test.go:2434: DockerSuite.TestRunMountShmMqueueFromHost

docker_cli_run_test.go:2445:
    out, _ := dockerCmd(c, "run", "--name", "ipchost", "--ipc", "host", "busybox", "cat", "/dev/shm/test")
/go/src/github.com/docker/docker/pkg/integration/dockerCmd_utils.go:35:
    c.Assert(err, check.IsNil, check.Commentf("%q failed with errors: %s, %v", strings.Join(args, " "), out, err))
... value *exec.ExitError = &exec.ExitError{ProcessState:(*os.ProcessState)(0xc820b12d40)} ("exit status 125")
... "run --name ipchost --ipc host busybox cat /dev/shm/te
---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-Win2Lin [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-Win2Lin/21540/console):

~~~console

---
ttachSilent	1.716s
PASS: docker_cli_start_test.go:112: DockerSuite.TestStartMultipleContainers	2.299s
PASS: docker_cli_start_test.go:96: DockerSuite.TestStartPausedContainer	0.850s
PASS: docker_cli_start_test.go:71: DockerSuite.TestStartRecordError	1.819s
PASS: docker_api_containers_test.go:1425: DockerSuite.TestStartWithNilDNS	0.688s
PASS: docker_api_containers_test.go:882: DockerSuite.TestStartWithTooLowMemoryLimit	0.141s

----------------------------------------------------------------------
FAIL: docker_cli_stats_test.go:98: DockerSuite.TestStatsAllNewContainersAdded

docker_cli_stats_test.go:131:
    c.Fatal("failed to observe new container created added to stats")
... Error: failed to observe new container created added to stats


----------------------------------------------------------------------
PASS: docker_cli_stats_test.go:80: DockerSuite.TestStatsAllNoStream	3.498s
PASS: docker_cli_stats_test.go:56: DockerSuite.TestStatsAllRunningNoStream	3.910s
PASS: docker_cli_stats_test.go:43: DockerSuite.TestStatsContainerNotFound	3.175s
PASS: docker_cli_stats
---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-gccgo [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-gccgo/1694/console):

~~~console

---
docker_integration_cli.DockerSuite	7.064s
PASS: docker_cli_run_test.go:2701: TestRunRestartMaxRetries.pN52_github_com_docker_docker_integration_cli.DockerSuite	3.985s
PASS: docker_cli_run_test.go:2147: TestRunReuseBindVolumeThatIsSymlink.pN52_github_com_docker_docker_integration_cli.DockerSuite	2.124s
PASS: docker_cli_run_test.go:1179: TestRunRootWorkdir.pN52_github_com_docker_docker_integration_cli.DockerSuite	1.091s

----------------------------------------------------------------------
FAIL: docker_cli_run_unix_test.go:842: TestRunSeccompAllowAptKey.pN52_github_com_docker_docker_integration_cli.DockerSuite

docker_cli_run_unix_test.go:848:
    c.Fatalf("expected apt-key with seccomp to succeed, got %s: %v", out, err)
... Error: expected apt-key with seccomp to succeed, got Executing: gpg --ignore-time-conflict --no-options --no-default-keyring --homedir /tmp/tmp.OBAzubhEAg --no-auto-check-trustdb --trust-model always --primary-keyring /etc/apt/trusted.gpg --keyring /etc/apt/trusted.gpg.d/debian-archive-jessie-automatic.gpg --keyring /etc/apt/trusted.gpg.d/debian-archive-jessie-security-automat
---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-WoW-TP4 [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-WoW-TP4/783/console):

~~~console

---
b0267a.
+ ec=125
+ set +x

----------------------------------
ERROR: Failed to build test binary
----------------------------------



-----------------------------------------------
ERROR: Failed with exitcode 125 at Tue Feb  9 16:33:24 CUT 2016.
-----------------------------------------------


INFO: Tidying up at end of run
INFO: Nuking /d/CI
INFO: Zapped successfully
INFO: End of cleanup
INFO: Ended at Tue Feb  9 16:33:25 CUT 2016 (1m 12s)
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[docker] $ sh -xe D:\temp\hudson6600222008511807087.sh
+ set +e
+ set +x
INFO: End of cleanup
Notifying endpoint 'HTTP:https://leeroy.dockerproject.org/notification/jenkins'
---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-WoW-TP4 [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-WoW-TP4/784/console):

~~~console

---
97081c.
+ ec=125
+ set +x

----------------------------------
ERROR: Failed to build test binary
----------------------------------



-----------------------------------------------
ERROR: Failed with exitcode 125 at Tue Feb  9 17:00:11 CUT 2016.
-----------------------------------------------


INFO: Tidying up at end of run
INFO: Nuking /d/CI
INFO: Zapped successfully
INFO: End of cleanup
INFO: Ended at Tue Feb  9 17:00:12 CUT 2016 (1m 12s)
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[docker] $ sh -xe D:\temp\hudson4304507035559340228.sh
+ set +e
+ set +x
INFO: End of cleanup
Notifying endpoint 'HTTP:https://leeroy.dockerproject.org/notification/jenkins'
---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-userns [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-userns/6028/console):

~~~console

---
pping root)
SKIP: docker_cli_run_test.go:3667: DockerSuite.TestRunModeNetContainerHostname (Test cannot be run when remapping root)
SKIP: docker_cli_run_test.go:2479: DockerSuite.TestRunModePidHost (Test cannot be run when remapping root)
PASS: docker_cli_run_test.go:2501: DockerSuite.TestRunModeUTSHost	1.603s
SKIP: docker_cli_run_test.go:2102: DockerSuite.TestRunMountOrdering (Test cannot be run when remapping root)

----------------------------------------------------------------------
FAIL: docker_cli_run_test.go:2434: DockerSuite.TestRunMountShmMqueueFromHost

docker_cli_run_test.go:2445:
    out, _ := dockerCmd(c, "run", "--name", "ipchost", "--ipc", "host", "busybox", "cat", "/dev/shm/test")
/go/src/github.com/docker/docker/pkg/integration/dockerCmd_utils.go:35:
    c.Assert(err, check.IsNil, check.Commentf("%q failed with errors: %s, %v", strings.Join(args, " "), out, err))
... value *exec.ExitError = &exec.ExitError{ProcessState:(*os.ProcessState)(0xc820bf0460)} ("exit status 125")
... "run --name ipchost --ipc host busybox cat /dev/shm/te
---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-userns [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-userns/6029/console):

~~~console

---
pping root)
SKIP: docker_cli_run_test.go:3667: DockerSuite.TestRunModeNetContainerHostname (Test cannot be run when remapping root)
SKIP: docker_cli_run_test.go:2479: DockerSuite.TestRunModePidHost (Test cannot be run when remapping root)
PASS: docker_cli_run_test.go:2501: DockerSuite.TestRunModeUTSHost	0.967s
SKIP: docker_cli_run_test.go:2102: DockerSuite.TestRunMountOrdering (Test cannot be run when remapping root)

----------------------------------------------------------------------
FAIL: docker_cli_run_test.go:2434: DockerSuite.TestRunMountShmMqueueFromHost

docker_cli_run_test.go:2445:
    out, _ := dockerCmd(c, "run", "--name", "ipchost", "--ipc", "host", "busybox", "cat", "/dev/shm/test")
/go/src/github.com/docker/docker/pkg/integration/dockerCmd_utils.go:35:
    c.Assert(err, check.IsNil, check.Commentf("%q failed with errors: %s, %v", strings.Join(args, " "), out, err))
... value *exec.ExitError = &exec.ExitError{ProcessState:(*os.ProcessState)(0xc821aa52a0)} ("exit status 125")
... "run --name ipchost --ipc host busybox cat /dev/shm/te
---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-WoW-TP4 [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-WoW-TP4/786/console):

~~~console

---
a7865e.
+ ec=125
+ set +x

----------------------------------
ERROR: Failed to build test binary
----------------------------------



-----------------------------------------------
ERROR: Failed with exitcode 125 at Tue Feb  9 17:40:40 CUT 2016.
-----------------------------------------------


INFO: Tidying up at end of run
INFO: Nuking /d/CI
INFO: Zapped successfully
INFO: End of cleanup
INFO: Ended at Tue Feb  9 17:40:41 CUT 2016 (1m 13s)
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[docker] $ sh -xe D:\temp\hudson8544444697795995595.sh
+ set +e
+ set +x
INFO: End of cleanup
Notifying endpoint 'HTTP:https://leeroy.dockerproject.org/notification/jenkins'
---

~~~

crosbymichael commented:
LGTM

GordonTheTurtle commented:
Job: Docker-PRs-Win2Lin [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-Win2Lin/21559/console):

~~~console

---
         125.1 MB
busybox             latest              d9551b4026f0        9 weeks ago         1.113 MB
hello-world         frozen              690ed74de00f        3 months ago        960 B
emptyfs             latest              11f64303f0f7        2 years ago         0 B


-----------------------------------------------
Windows CI Timeout Detection - End of debugging
-----------------------------------------------


c:\scripts>exit 0 
POST BUILD TASK : SUCCESS
END OF POST BUILD TASK : 0
[PostBuildScript] - Execution post build scripts.
[docker] $ sh -xe C:\Users\Jenkins\AppData\Local\Temp\hudson3176938035746148017.sh
+ set +x


---

~~~

mlaventure committed:
Disable TestRunMountshmmqueuefromhost when using userns

Since we now automatically mount the mqueue device inside the
container (instead of bind mounting the one from the host), when
trying to start a container with --ipc=host, the mount will fail with
EPERM.

Signed-off-by: Kenfe-Mickael Laventure <mickael.laventure@gmail.com>

GordonTheTurtle commented:
Job: Docker-PRs-WoW-TP4 [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-WoW-TP4/794/console):

~~~console

---
469117.
+ ec=125
+ set +x

----------------------------------
ERROR: Failed to build test binary
----------------------------------



-----------------------------------------------
ERROR: Failed with exitcode 125 at Tue Feb  9 19:32:37 CUT 2016.
-----------------------------------------------


INFO: Tidying up at end of run
INFO: Nuking /d/CI
INFO: Zapped successfully
INFO: End of cleanup
INFO: Ended at Tue Feb  9 19:32:38 CUT 2016 (1m 14s)
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[docker] $ sh -xe D:\temp\hudson7295603797677312460.sh
+ set +e
+ set +x
INFO: End of cleanup
Notifying endpoint 'HTTP:https://leeroy.dockerproject.org/notification/jenkins'
---

~~~

cpuguy83 commented:
Looks like you have a couple of extra commits

rhatdan commented:
LGTM

mlaventure commented:
@cpuguy83 thanks, fixed.

jfrazelle commented:
LGTM

GordonTheTurtle commented:
Job: Docker-PRs-WoW-TP4 [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-WoW-TP4/800/console):

~~~console

---
cd580c.
+ ec=125
+ set +x

----------------------------------
ERROR: Failed to build test binary
----------------------------------



-----------------------------------------------
ERROR: Failed with exitcode 125 at Tue Feb  9 21:38:53 CUT 2016.
-----------------------------------------------


INFO: Tidying up at end of run
INFO: Nuking /d/CI
INFO: Zapped successfully
INFO: End of cleanup
INFO: Ended at Tue Feb  9 21:38:54 CUT 2016 (1m 13s)
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[docker] $ sh -xe D:\temp\hudson7516330331215967345.sh
+ set +e
+ set +x
INFO: End of cleanup
Notifying endpoint 'HTTP:https://leeroy.dockerproject.org/notification/jenkins'
---

~~~

cpuguy83 commented on integration-cli/check_test.go:
```diff
@@ -33,6 +33,7 @@ func (s *DockerSuite) TearDownTest(c *check.C) {
 	deleteAllImages()
 	deleteAllVolumes()
 	deleteAllNetworks()
+	cleanupShm()
```
Can we just run this on the test that needs it?

mlaventure replied:
Couldn't figure out a way to do so. If the previous test crashes this is the only certain way to clean up behind that I found. As far as CI is concerned this could be safely removed though since we start from scratch every time. It just means that after a test failure there is no warranty that the test actually passes if it also uses shm.


Any suggestions?

cpuguy83 replied:
Did you try with just a defer at the top of the test?

mlaventure replied:
That was my first instinct, but os.Exit doesn't honor defer

mlaventure replied:
My bad, just checked go-check code.  It actually handles defer. I'll update my patch.

cpuguy83 replied:
yay :+1:

GordonTheTurtle commented:
Job: Docker-PRs-Win2Lin [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-Win2Lin/21571/console):

~~~console

---
aking bundle: binary (in bundles/1.11.0-dev/binary)
Building: bundles/1.11.0-dev/binary/docker-1.11.0-dev
# github.com/docker/docker/docker
/usr/local/go/pkg/tool/linux_amd64/link: running gcc failed: fork/exec /usr/bin/gcc: cannot allocate memory

---
Build step 'Execute shell' marked build as failure
Performing Post build task...
Could not match :Build timed out  : False
Logical operation result is FALSE
Skipping script  : cd c:\scripts
sh timeout-post-build.sh

END OF POST BUILD TASK 	: 0
[PostBuildScript] - Execution post build scripts.
[docker] $ sh -xe C:\Users\Jenkins\AppData\Local\Temp\hudson780522057342652313.sh
+ set +x


---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-WoW-TP4 [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-WoW-TP4/802/console):

~~~console

---
1ecd24.
+ ec=125
+ set +x

----------------------------------
ERROR: Failed to build test binary
----------------------------------



-----------------------------------------------
ERROR: Failed with exitcode 125 at Tue Feb  9 22:18:52 CUT 2016.
-----------------------------------------------


INFO: Tidying up at end of run
INFO: Nuking /d/CI
INFO: Zapped successfully
INFO: End of cleanup
INFO: Ended at Tue Feb  9 22:18:53 CUT 2016 (1m 16s)
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[docker] $ sh -xe D:\temp\hudson170925096743446905.sh
+ set +e
+ set +x
INFO: End of cleanup
Notifying endpoint 'HTTP:https://leeroy.dockerproject.org/notification/jenkins'
---

~~~

cpuguy83 commented:
LGTM pending janky.
